File: src/main/java/no/nav/syfo/services/GeografiskTilgangService.java
package no.nav.syfo.services;

import no.nav.syfo.axsys.AxsysConsumer;
import no.nav.syfo.axsys.AxsysEnhet;
import no.nav.syfo.behandlendeenhet.BehandlendeEnhetConsumer;
import no.nav.syfo.domain.AdRoller;
import no.nav.syfo.norg2.NorgConsumer;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.List;

import static java.util.stream.Collectors.toList;

@Service
public class GeografiskTilgangService {

    private final AxsysConsumer axsysConsumer;
    private final BehandlendeEnhetConsumer behandlendeEnhetConsumer;
    private final LdapService ldapService;
    private final NorgConsumer norgConsumer;

    @Autowired
    public GeografiskTilgangService(
            AxsysConsumer axsysConsumer,
            BehandlendeEnhetConsumer behandlendeEnhetConsumer,
            LdapService ldapService,
            NorgConsumer norgConsumer
    ) {
        this.axsysConsumer = axsysConsumer;
        this.behandlendeEnhetConsumer = behandlendeEnhetConsumer;
        this.ldapService = ldapService;
        this.norgConsumer = norgConsumer;
    }

    public boolean harGeografiskTilgang(String veilederId, String personFnr, String geografiskTilknytning) {
        if (harNasjonalTilgang(veilederId)) {
            return true;
        }
        final String navKontorForGT = getNavKontorForGT(personFnr, geografiskTilknytning);
        final List<String> veiledersEnheter = axsysConsumer.enheter(veilederId)
                .stream()
                .map(AxsysEnhet::getEnhetId)
                .collect(toList());

        return harLokalTilgangTilBrukersEnhet(navKontorForGT, veiledersEnheter)
                || harRegionalTilgangTilBrukersEnhet(navKontorForGT, veiledersEnheter, veilederId);
    }

    private boolean harNasjonalTilgang(String veilederId) {
        return ldapService.harTilgang(veilederId, AdRoller.NASJONAL.rolle)
                || ldapService.harTilgang(veilederId, AdRoller.UTVIDBAR_TIL_NASJONAL.rolle);
    }

    private boolean harLokalTilgangTilBrukersEnhet(String navKontorForGT, List<String> veiledersEnheter) {
        return veiledersEnheter.contains(navKontorForGT);
    }

    private boolean harRegionalTilgang(String veilederId) {
        return ldapService.harTilgang(veilederId, AdRoller.REGIONAL.rolle)
                || ldapService.harTilgang(veilederId, AdRoller.UTVIDBAR_TIL_REGIONAL.rolle);
    }

    private boolean harRegionalTilgangTilBrukersEnhet(String navKontorForGT, List<String> veiledersEnheter, String veilederId) {
        List<String> veiledersOverordnedeEnheter = veiledersEnheter.stream()
                .map(norgConsumer::getOverordnetEnhetListForNAVKontor)
                .flatMap(Collection::stream)
                .collect(toList());

        return harRegionalTilgang(veilederId) && norgConsumer.getOverordnetEnhetListForNAVKontor(navKontorForGT)
                .stream()
                .anyMatch(veiledersOverordnedeEnheter::contains);
    }

    private String getNavKontorForGT(String personFnr, String geografiskTilknytning) {
        return isGeografiskTilknytningUtland(geografiskTilknytning)
                ? behandlendeEnhetConsumer.getBehandlendeEnhet(personFnr, null).getEnhetId()
                : norgConsumer.getNAVKontorForGT(geografiskTilknytning);
    }

    private boolean isGeografiskTilknytningUtland(String geografiskTilknytning) {
        return geografiskTilknytning.matches("[a-zA-Z]{3}");
    }
}

File: src/main/kotlin/no/nav/syfo/geografisktilknytning/GeografiskTilgangService.kt
package no.nav.syfo.geografisktilknytning

import no.nav.syfo.consumer.axsys.AxsysConsumer
import no.nav.syfo.consumer.axsys.AxsysEnhet
import no.nav.syfo.consumer.behandlendeenhet.BehandlendeEnhetConsumer
import no.nav.syfo.domain.AdRoller
import no.nav.syfo.consumer.norg2.NorgConsumer
import no.nav.syfo.consumer.pdl.PdlConsumer
import no.nav.syfo.consumer.ldap.LdapService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.stream.Collectors

@Service
class GeografiskTilgangService @Autowired constructor(
    private val axsysConsumer: AxsysConsumer,
    private val behandlendeEnhetConsumer: BehandlendeEnhetConsumer,
    private val ldapService: LdapService,
    private val norgConsumer: NorgConsumer,
    private val pdlConsumer: PdlConsumer
) {
    fun harGeografiskTilgang(veilederId: String, personFnr: String): Boolean {
        if (harNasjonalTilgang(veilederId)) {
            return true
        }
        val geografiskTilknytning = pdlConsumer.geografiskTilknytning(personFnr)
        val navKontorForGT = getNavKontorForGT(personFnr, geografiskTilknytning)
        val veiledersEnheter = axsysConsumer.enheter(veilederId)
            .stream()
            .map(AxsysEnhet::enhetId)
            .collect(Collectors.toList())
        return (harLokalTilgangTilBrukersEnhet(navKontorForGT, veiledersEnheter)
            || harRegionalTilgangTilBrukersEnhet(navKontorForGT, veiledersEnheter, veilederId))
    }

    private fun harNasjonalTilgang(veilederId: String): Boolean {
        return (ldapService.harTilgang(veilederId, AdRoller.NASJONAL.rolle)
            || ldapService.harTilgang(veilederId, AdRoller.UTVIDBAR_TIL_NASJONAL.rolle))
    }

    private fun harLokalTilgangTilBrukersEnhet(navKontorForGT: String, veiledersEnheter: List<String>): Boolean {
        return veiledersEnheter.contains(navKontorForGT)
    }

    private fun harRegionalTilgang(veilederId: String): Boolean {
        return (ldapService.harTilgang(veilederId, AdRoller.REGIONAL.rolle)
            || ldapService.harTilgang(veilederId, AdRoller.UTVIDBAR_TIL_REGIONAL.rolle))
    }

    private fun harRegionalTilgangTilBrukersEnhet(navKontorForGT: String, veiledersEnheter: List<String>, veilederId: String): Boolean {
        val veiledersOverordnedeEnheter = veiledersEnheter.map { enhetNr: String ->
            norgConsumer.getOverordnetEnhetListForNAVKontor(enhetNr)
        }.flatten()

        return harRegionalTilgang(veilederId) && norgConsumer.getOverordnetEnhetListForNAVKontor(navKontorForGT)
            .any { overordnetEnhet: String -> veiledersOverordnedeEnheter.contains(overordnetEnhet) }
    }

    private fun getNavKontorForGT(personFnr: String, geografiskTilknytning: String): String {
        return if (isGeografiskTilknytningUtland(geografiskTilknytning)) behandlendeEnhetConsumer.getBehandlendeEnhet(personFnr, null).enhetId else norgConsumer.getNAVKontorForGT(geografiskTilknytning)
    }

    private fun isGeografiskTilknytningUtland(geografiskTilknytning: String): Boolean {
        return geografiskTilknytning.matches(Regex("[a-zA-Z]{3}"))
    }
}
